[GlobalISel][AArch64][AMDGPU] Expand FPOWI into series of multiplication

isuckatcs opened the pull request:
SelectionDAG already converts FPOWI into multiplications, this patch introduces the same optimization into GlobalISel.

tschuett commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7142,14 +7142,50 @@ LegalizerHelper::lowerFPTRUNC(MachineInstr &MI) {
+    ExpVal = -ExpVal;
+
+  Register Res = MRI.createGenericVirtualRegister(Ty);
+  MIRBuilder.buildCopy(Res, Base);
```
```
auto Copy = MIRBuilder.buildCopy(Ty, Base);
```

isuckatcs replied:
I only see the following definition
```c++
MachineInstrBuilder MachineIRBuilder::buildCopy(const DstOp &Res,
                                                const SrcOp &Op) {
  return buildInstr(TargetOpcode::COPY, Res, Op);
}
```
Am I missing something here?

tschuett replied:
No you don't. Scroll up to `lowerFPTRUNC_F64_TO_F16`. You will see the same pattern.

tschuett replied:
It looks cuter, when the first parameter is a type. Then you don't need to create virtual registers manually.

isuckatcs replied:
I agree, this patterns is awesome. Thanks for the heads-up!

tschuett commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7142,14 +7142,50 @@ LegalizerHelper::lowerFPTRUNC(MachineInstr &MI) {
+
+  while (--ExpVal > 0) {
+    Register Tmp = MRI.createGenericVirtualRegister(Ty);
+    MIRBuilder.buildFMul(Tmp, Res, Base);
```
```
auto Tmp = MIRBuilder.buildFMul(Ty, Res, Base);
```

arsenm commented:
> SelectionDAG already converts FPOWI into multiplications, this patch introduces the same optimization into GlobalISel.

Where do you see this optimization? I'm pretty sure this is not a viable lowering strategy. It would require some set of unsafe math flags, and should go into a combiner, not the legalizer

isuckatcs commented:
> Where do you see this optimization?

It is in [SelectionDAGBuilder.cpp](https://github.com/isuckatcs/llvm-project/blob/2ca8c856eeae739ec1e7242ee7e69f99ecf376d3/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L5790)

arsenm commented:
> > Where do you see this optimization?
> 
> It is in [SelectionDAGBuilder.cpp](https://github.com/isuckatcs/llvm-project/blob/2ca8c856eeae739ec1e7242ee7e69f99ecf376d3/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L5790)

Ugh, the DAG builder should definitely not be doing optimizations like this. But yes, this only works on constants and thus should not be a legalization strategy. It should be an optimization only in the CombinerHelper

tschuett commented:
```
def expand_fpowi : GICombineRule<
   (defs root:$root, build_fn_matchinfo:$matchinfo),
   (match (G_CONSTANT $int, $imm)
          (G_FPOWI $root, $float, $int),
   [{ return Helper.matchExpandFPOWI(*${root}, ${matchinfo}); }]),
   (apply [{ Helper.applyBuildFnBO(*${root}, ${matchinfo}); }])>;
```

aemerson commented:
> > > Where do you see this optimization?
> > 
> > 
> > It is in [SelectionDAGBuilder.cpp](https://github.com/isuckatcs/llvm-project/blob/2ca8c856eeae739ec1e7242ee7e69f99ecf376d3/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L5790)
> 
> Ugh, the DAG builder should definitely not be doing optimizations like this. But yes, this only works on constants and thus should not be a legalization strategy. It should be an optimization only in the CombinerHelper

+1 please move this to the combiner.

isuckatcs committed:
[GlobalISel][AArch64][AMDGPU] Expand FPOWI into series of multiplication

SelectionDAG already converts FPOWI into multiplications, this patch
introduces the same optimization into GlobalISel.

isuckatcs commented:
I moved the optimization to the combiner as requested.

tschuett commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,38 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+
+  int64_t OriginalExprVal = *ConstantExpValue;
+  int64_t ExpVal = OriginalExprVal;
+
```
```
if (!MRI.hasOneNonDBGUse(Base))
  return false;
```

isuckatcs replied:
Why is this needed? Why can't we optimize if there are multiple uses of Base?

isuckatcs replied:
@tschuett sorry for the ping, but can you please help me understand why this change is needed, so I can push the PR to completion?

tschuett replied:
Ignore that comment.

tschuett commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,38 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+  auto [Dst, Base, Exp] = MI.getFirst3Regs();
+  LLT Ty = MRI.getType(Dst);
+
+  MachineRegisterInfo &MRI = *Builder.getMRI();
```
`MRI` is a local variable.

tschuett commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,38 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
   Root->eraseFromParent();
 }
 
+void CombinerHelper::applyExpandFPowI(MachineInstr &MI, BuildFnTy &MatchInfo) {
```
`MatchInfo` is dead.

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,38 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+  if (ExpVal < 0)
+    ExpVal = -ExpVal;
+
+  auto Res = Builder.buildCopy(Ty, Base);
```
Why do you need this copy? You can just initialize this to Base?

arsenm commented on llvm/test/CodeGen/AArch64/GlobalISel/combine-fpowi.mir:
```diff
@@ -0,0 +1,127 @@
+      %1:_(s32) = COPY $w0
+      %2:_(s64) = G_FPOWI %0, %1(s32)
+      $d0 = COPY %2(s64)
+...
```
Add a test for flag behavior

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,38 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+
+  // If the original was negative, invert the result, producing 1/(x*x*x).
+  if (OriginalExprVal < 0)
+    Res = Builder.buildFDiv(Ty, Builder.buildFConstant(Ty, 1.0), Res);
```
You are losing flags on the original instruction

isuckatcs commented:
I addressed all the comments on the PR. Is there anything else I could do to improve it, or can we merge it in this state?

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,38 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+  auto [Dst, Base, Exp] = MI.getFirst3Regs();
+  LLT Ty = MRI.getType(Dst);
+
+  std::optional<int64_t> ConstantExpValue = getIConstantVRegSExtVal(Exp, MRI);
```
Can you avoid the lookup again? Already matched the immediate in the pattern

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,38 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+
+  if (ExpVal < 0)
+    ExpVal = -ExpVal;
+
```
Can you preserve the comment from the DAG version about "We use the simple binary decomposition method..."

arsenm reviewed (commented):
Basically lgtm

davemgreen reviewed (commented):
Should this call isBeneficialToExpandPowI too? Otherwise it might expand to a lot of multiplies.

isuckatcs commented:
> Should this call isBeneficialToExpandPowI too? Otherwise it might expand to a lot of multiplies.

Said function is defined like this:
```c++
  /// Return true if it is beneficial to expand an @llvm.powi.* intrinsic.
  /// If not optimizing for size, expanding @llvm.powi.* intrinsics is always
  /// considered beneficial.
  /// If optimizing for size, expansion is only considered beneficial for upto
  /// 5 multiplies and a divide (if the exponent is negative).
  bool isBeneficialToExpandPowI(int64_t Exponent, bool OptForSize) const {
    if (Exponent < 0)
      Exponent = -Exponent;
    uint64_t E = static_cast<uint64_t>(Exponent);
    return !OptForSize || (llvm::popcount(E) + Log2_64(E) < 7);
  }
```
So, the intrinsic is always expanded unless if we optimize for size. I haven't seen any check
for size optimization inside selection dag, and I'm not even sure we check it there.

arsenm commented:
There certainly are some size checks in the DAG, if not many. Corresponds to optsize attribute

isuckatcs commented:
Sorry, I made a typo in my previous comment. I meant that I haven't seen many checks for size optimization in GlobalISel. I added one regardless. 

Also it turned out I accidentally expanded G_FPOWI into a lot of multiplications instead of using the more optimal binary exponentiation algorithm, which is also used in the DAG and results in fewer multiplications.

davemgreen commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -7356,6 +7356,59 @@ void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+// consider moving the two to some common place.
+bool CombinerHelper::matchFPowIExpansion(MachineInstr &MI, int64_t Exponent) {
+  bool OptForSize = MI.getMF()->getFunction().hasOptSize();
+  if (Exponent < 0)
+    Exponent = -Exponent;
+  uint64_t E = static_cast<uint64_t>(Exponent);
+  return !OptForSize || (llvm::popcount(E) + Log2_64(E) < 7);
```
It should be able to call `getTargetLowering().isBeneficialToExpandPowI(Exponent, OptForSize);`. In case it changes in the future we can make sure they stay in sync.

davemgreen reviewed (commented):
Thanks. I had actually misread isBeneficialToExpandPowI and hadn't realized the heuristic was OptSize only - that the tree of multiplies was efficient enough for any possible constant. That's cool it works like that.

Other than calling isBeneficialToExpandPowI directly this looks OK to me.

davemgreen reviewed (approved):
Thanks. LGTM if @arsenm agrees.
